doc/vlan_stacking/vlan_stacking_HLD.md

```diff
@@ -0,0 +1,708 @@
+  - [CLI/YANG model Enhancements](#cliyang-model-enhancements)
+    - [CLI](#cli)
+    - [Yang model](#yang-model)
+  - [Examples](#examples)
+    - [Examples of QinQ Configuration](#examples-of-qinq-configuration)
+    - [Examples of VLAN Translation Configuration](#examples-of-vlan-translation-configuration)
+- [Warmboot and Fastboot Design Impact](#warmboot-and-fastboot-design-impact)
+- [Restrictions/Limitations](#restrictionslimitations)
+- [Testing Requirements/Design](#testing-requirementsdesign)
+  - [System Test Cases](#system-test-cases)
+- [Future Work](#future-work)
+
+# List of Tables
+
+* [Table 1: Revision](#revision)
+* [Table 2: Definitions/Abbreviations](#definitionsabbreviations)
+* [Table 3: SAI attributes related to VLAN stacking](#sai-api)
+
+# Revision
+
+###### Table 1: Revision
+
+| Rev   |   Date     | Author        | Change Description |
+|:-----:|:----------:|:-------------:|--------------------|
+| 0.1   |20-Dec-2021 | Tommy Tseng   | Initial version    |
+| 0.2   |26-Jan-2022 | Tommy Tseng   | Update CLI command |
+# Scope
+
+This document describes the high level design of VLAN stacking feature.
+
+# Definitions/Abbreviations
+
+###### Table 2: Definitions/Abbreviations
+
+| Abbreviation | Full form              |
+|--------------|------------------------|
+| VID          | Virtual LAN Identifier |
+| S-VLAN       | Service Provider VLAN  |
+| C-VLAN       | Customer VLAN          |
+| S-TAG        | S-VLAN tag             |
+| C-TAG        | C-VLAN tag             |
+
+# Overview
+
+This document describes the design details of the VLAN stacking feature. VLAN stacking is a feature designed for service providers who carry traffic of multiple customers across  their networks and are required to maintain the VLAN and Layer 2 protocol configurations of each customer without impacting the traffic of other customers. VLAN stacking include QinQ feature and VLAN Translation feature.
+
+IEEE 802.1Q tunneling (QinQ tunneling) uses a single Service Provider VLAN (SPVLAN) for customers who have multiple VLANs. Customer VLAN IDs are preserved and traffic from different customers is segregated within the service provider’s network even when they use the same customer-specific VLAN IDs. QinQ tunneling expands VLAN space by using a VLAN-in-VLAN hierarchy, preserving the customer’s original tagged packets, and adding SPVLAN tags to each frame (also called double tagging).
+
+QinQ tunneling uses double tagging to preserve the customer’s VLAN tags on traffic crossing the service provider’s network. However, if any switch in the path crossing the service provider’s network does not support this feature, then the switches directly connected to that device can be configured to swap the customer’s VLAN ID with the service provider’s VLAN ID for upstream traffic, or the service provider’s VLAN ID with the customer’s VLAN ID for downstream traffic.
+
+## VLAN Stacking Deployment Use Cases
+
```

Comment:
Hello @pttch
Here are our comments from Dell team (Senthil Kumar Ganesan, Vinoth Kumar Arumugam, and myself)

Changes in the config-DB:
1) Enhance PORT & PORTCHANNEL table to include the fields 
      a) customer edge port (true/false) 
      b) SVID field (valid only if a) is set) 
      c) SVID override priority (valid only if a) is set) -> (if this field is not configured, the default behavior is to inherit 
          priority from C-tag) 
   -> Upon getting any C-tag on this port/port-channel, SVID should be pushed on the ingress stage and the same be popped on the egress stage.

2) VLAN_STACKING -> Keys: a) Port/Portchannel b) SVID 
                                     fields: cvid leaf-list
       -> This table config would be used when there is a need for explicit CVID to SVID and vice-versa mapping

3) VLAN_XLATE -> This table will be used to translate from one VLAN to another, there can be a field to indicate whether the translation is expected on the inner or outer VLAN tag.

Reply:
@venkatmahalingam Thanks for your suggestion, Could you share the mib files to me ?

Reply:
Sure, I sent it to your email.

Reply:
@venkatmahalingam 
After internal discuss, we have the questions, 
1. a) customer edge port (true/false) --> does this only use in VLAN_Stacking, because VLAN_XLATE is in another table.
2. b) SVID field (valid only if a) is set) --> what's use for ?  It shall cover in VLAN_STACKING table.
3. c) SVID override priority (valid only if a) is set) --> what's use for ?  It shall cover in VLAN_STACKING table.

Reply:
> @venkatmahalingam After internal discuss, we have the questions,
> 
> 1. a) customer edge port (true/false) --> does this only use in VLAN_Stacking, because VLAN_XLATE is in another table.
> 2. b) SVID field (valid only if a) is set) --> what's use for ?  It shall cover in VLAN_STACKING table.
> 3. c) SVID override priority (valid only if a) is set) --> what's use for ?  It shall cover in VLAN_STACKING table.

VLAN stacking can be done in two ways,

1) Upon getting any CVID on the port/port-channel (customer edge port), push SVID in the ingress stage, and on the egress stage, SVID should be popped, if we don't want to inherit CVID prioirty, SVID priority should be configured explicitly.

2) VLAN_STACKING is an alternative method to do the SVID mapping for a given [cvid list, port/port-channel]
   VLAN_STACKING -> Keys: 1) Port/Portchannel 2) SVID 
                                     fields: cvid leaf-list

If you have any further questions, we can discuss in the meeting.

Reply:
@venkatmahalingam 
So we may have the following design change, it shall include first behavior
VLAN_STACKING:
  key: port + svid
  svid-priority: xx(0~7: if we want to override, -1: inherit from CVID priority)
  cvid-list: [...] 
If you feel well for this, we will start to re-define the db. Thanks.

Reply:
Yes, please go ahead and update the HLD for review again.

Reply:
@venkatmahalingam 
Regarding to the VLAN_Transition we will use the following table. 
VLAN_TRANSLATION:
key: port + svid
cvid: int
If you feel well for this, we will also re-define the db.

Reply:
@venkatmahalingam Hi, I'm update the DB schema, please have a look for this. If this is ok, please also help to approve this.

Reply:
@pttch I don't see the below config option in the HLD, this will satisfy the requirement for adding a s-tag for traffic with any c-tag coming on the port/port-channel, please let us know if you have any plan to take care of this as well in this HLD.

Enhance PORT & PORTCHANNEL table to include the fields
a) customer edge port (true/false)
b) SVID field (valid only if a) is set)
c) SVID override priority (valid only if a) is set) -> (if this field is not configured, the default behavior is to
inherit priority from C-tag)
-> Upon getting any C-tag on this port/port-channel, SVID should be pushed on the ingress stage and the same should be popped on the egress stage.

Reply:
@venkatmahalingam 
a) customer edge port (true/false)
b) SVID field (valid only if a) is set)
--> It can use vlan range setting for all c-vlan, this example show all c-vlan will push s-vlan 100 in Ethernet1 and strips S-VLAN 100 in egress side.
"VLAN_STACKING": {
    "Ethernet1|100": {
        "c_vlanids": ["1..4095"],
        "s_vlan_priority": "0"
    }
} 
c) SVID override priority (valid only if a) is set) -> (if this field is not configured, the default behavior is to
inherit priority from C-tag)
--> Yes, the sai do not implement this, we will do this in the future.